DOC add trap definitions (#809)
fairlearn/fairlearn: docs/user_guide/fairness_in_machine_learning.rst

subpackage :code:`fairlearn.metrics`.

Failure modes encountered when abstracting
--------------------------------------------------

Machine learning systems used in the real world are inherently sociotechnical
systems, which include both technologies and social actors. Designers of machine
learning systems who fail to consider the way in which social contexts and technologies
are interrelated are at risk of falling into "abstraction traps". In this section,
we explain what those traps are, and give some suggestions on how you can avoid them.

In "Fairness and Abstraction in Sociotechnical Systems" Selbst et al. [#4]_
identify failure modes that arise from applying abstraction. They identify
them as:

* *The Solutionism Trap*

* *The Ripple Effect Trap*

* *The Formalism Trap*

* *The Portability Trap*

* *The Framing Trap*

These traps are inspired by the observation that each of these traps are the
result of failing to consider the way in with social context and technology
are interrelated, as well as a deeper understanding of "the social" in order to
solve problems Ackerman [#5]_.

The Solutionism Trap
^^^^^^^^^^^^^^^^^^^^

> This trap occurs when data scientists assume that the best solution to a problem involves

Reviewer: I am a bit torn between (1) speaking to our primary audience (data scientists), (2) acknowledging that data scientists usually don't (and probably shouldn't) make these decisions in isolation (e.g., influenced by management, product owners, project managers, etc.)

WDYT @mmadaio @lgfunderburk?

Author: yeah, I caught myself thinking something along these lines
Data scientists are those who implement models, but data scientists are not necessarily in charge of collecting data or creating data collections for the model

Author: Open to options, can change into "we" or "one"
When one assumes that the best solution involves....
Would this work better @hildeweerts ?

Author: changed it to we

technology, and fail to recognize other possible solutions outside of this realm.

> One area where this manifests in is within contexts in which the definition "fairness"

Reviewer: I am not sure if I understand what you mean with this sentence. Isn't the definition of "fairness" always dynamic / dependent on a political context? The sentence is also a bit hard to parse.

Author: Reworded as

`This trap occurs when we assume that the best solution to a problem
may involve technology, and fail to recognize other possible solutions outside of
this realm. This trap also occurs as definitions of fairness are
contested and changed and models become obsolete.`

Author: Reworded, and incorporated "fairness" as dynamic and dependent on a political context.

changes or is dependent on a political context.

Example: consider the problem of internet connectivity in rural communities. An example
of the solutionism trap is assuming that by using data science to study internet speed in
a given region, policies and deals can be arranged to reduce the price of internet access.
However, if there are additional socioeconomic problems within a community, for example lack of
education, infrastructure, information technology and health services, then an
algorithmic solution purely focused on internet speed will fail to meaningfully address
the needs of the community.

The Ripple Effect Trap
^^^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to understand how the insertion of technology into
an existing social system changes the behaviors and embedded values of the pre-existing system".

Example 1:

The Formalism Trap
^^^^^^^^^^^^^^^^^^

> This trap occurs when implementing an algorithmic solution failing to account the

Reviewer: I think this explanation can be further clarified by adding an introductory sentence on what is meant with formalism. E.g., something like "Many tasks of a data scientist involve some form of formalization: from measuring real-world phenomena as data to translating business KPI's and constraints into metrics, loss functions, or parameters. We fall into the formalism trap when this translation fails to account the social dimensions associated to the problem."

I personally always like to make the connection between the formalism trap and issues with construct validity, which I think are closely related.

Author: added a few more examples

Did go a little bit more into depth here, open to feedback

Author: Incorporated your introduction @hildeweerts

social dimensions associated to a problem. These dimensions include processes involved,
whether human, political or bureaucratic, as well as cultural and time-bound social
contexts. Such dimensions can often not be resolved via mathematical means.

> Failure occurs when steps are not taken to eradicate the limitations behind mathematical

Reviewer: This sentence might be a bit vague for people who are unfamiliar with what these limitations are exactly.

Author: agreed, modified quite a bit the formalism section, hopefully it is a bit more clear now

and statistical models.

Example: 

The Portability Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when data scientists fail to understand how reusing a model or algorithm
that is designed for one specific social context, may not necessarily apply to a different social
context. Reusing an algorithmic solution and failing to take into account differences in
involved social contexts can result in misleading results and potentially harmful consequences
if the algorithm is used to determine the fate of an individual.

Example: Reusing a machine learning algorithm used to screen job applications in the nursing
industry, for job applications in the information technology sector. An intuitive yet important

> difference between both context is the difference in skills required to succeed in both industries.

Reviewer: ```suggestion
difference between both contexts is the difference in skills required to succeed in both industries.
```

A slightly more subtle difference is the demographic differences in genders attracted to each
context, resulting from wording in job postings, social constructs and the male-female ratio
of successful applicants in each field.

The Framing Trap
^^^^^^^^^^^^^^^^

This trap occurs when data scientists fail to consider the full picture surrounding
a particular social context when designing and implementing a machine learning model

> in which the outcome involves enforcing decisions that will impact a person or group of people.

Reviewer: I think we can emphasize here that the narrow framing could exclude other technological systems but particularly social actors that influence the outcome of the system, in particular the system's users.

Author: Solved by incorporating your suggested paragraph @hildeweerts

Author: Added frames to emphasize the importance of incorporating both techncological and social systems

between the maximum and the minimum values. For more information refer to the
subpackage :code:`fairlearn.metrics`.

What traps can we fall into when abstracting a social problem?
--------------------------------------------------------------

Machine learning systems used in the real world are inherently sociotechnical
systems, which include both technologies and social actors. Designers of machine
learning systems who fail to consider the way in which social contexts and technologies
are interrelated are at risk of falling into "abstraction traps". In this section,
we explain what those traps are, and give some suggestions on how you can avoid them.

In "Fairness and Abstraction in Sociotechnical Systems" Selbst et al. [#4]_
identify failure modes that arise from applying abstraction. They identify
them as:

* *The Solutionism Trap*

* *The Ripple Effect Trap*

* *The Formalism Trap*

* *The Portability Trap*

* *The Framing Trap*

Each of these traps are the result of failing to consider the way in
with social context and technology are interrelated, as well as a deeper understanding of

> "the social" in order to solve problems Ackerman [#5]_.

Reviewer 2: I also wonder whether we want to include (some paraphrased, de-jargoned version of) the high-level summary of them from the takeaways section of the paper? Maybe here? Not sure - what do you think @lgfunderburk @hildeweerts 

[copy-pasted from the paper, but we'd want a less jargon-y version, probably]
(1) is appropriate to the situation in the first place, which requires a nuanced understanding of the relevant social context and its politics (Solutionism);
(2) affects the social context in a predictable way such that the problem that the technology solves remains unchanged after its introduction (Ripple Effect);
(3) can appropriately handle robust understandings of social requirements such as fairness, including the need for procedurality, contextuality, and contestability (Formalism);
(4) has appropriately modeled the social and technical requirements of the actual context in which it will be deployed (Portability); and
(5) is heterogeneously framed so as to include the data and social actors relevant to the localized question of fairness (Framing)

> "the social" in order to solve problems Ackerman [#5]_.

Reviewer: Great idea @mmadaio!

> "the social" in order to solve problems Ackerman [#5]_.

Author: Removed original sentences from paper, and added less jargon-y versions.

> "the social" in order to solve problems Ackerman [#5]_.

Author: Added sentence you provided

`We provide some definitions and examples of these traps to help Fairlearn users think about how choices they make in their work can avoid these common pitfalls.`

The Solutionism Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when we assume that the best solution to a problem
involves technology, and fail to recognize other possible solutions outside of
this realm. One area where this manifests in is within contexts in which the

> definition "fairness" changes or is dependent on a political context.

Author: Added @hildeweerts's sentence, modified to reflect that fairness is context dependent and contested

Example: consider the problem of internet connectivity in rural communities.
An example of the solutionism trap is assuming that by using data science to

> study internet speed in a given region, policies and deals can be arranged to

Author: Added this link https://fcm.ca/sites/default/files/documents/resources/report/broadband-access-rural-canada.pdf

Would it make more sense to list it as a reference?

> study internet speed in a given region, policies and deals can be arranged to

Author: Removed

reduce the price of internet access. However, if there are additional socioeconomic
problems within a community, for example lack of education, infrastructure, information
technology and health services, then an algorithmic solution purely focused on internet
speed will fail to meaningfully address the needs of the community.

The Ripple Effect Trap
^^^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to understand how the insertion of
technology into an existing social system changes the behaviors and embedded values
of the pre-existing system".

Example 1:

The Formalism Trap
^^^^^^^^^^^^^^^^^^

This trap occurs when implementing an algorithmic solution that fails to take

> into account the social dimensions associated to fairness in a situation. These

Author: Added @hildeweerts's examples and paragraph to motivate this in a more concrete way

Also expanded a bit more this section

dimensions include "procedurality", "contextuality" and "contestability". Such

> dimensions can often not be resolved via a purely mathematical framework.

Reviewer 2: Maybe something that gets at _why_ they can't be resolved, like:

"Because different definitions of fairness cannot be satisfied at the same time (e.g., [Chouldechova et al., 2017](https://arxiv.org/pdf/1703.00056.pdf), [although feel free to replace with a better citation if people have other preferences]), you'll need to decide which definition to use. But there's no mathematical way to make that decision. And it might be that _none_ of the definitions are appropriate for your situation.

Some reasons why people fall into this trap is because fairness is contextually-dependent, open to contestation by different groups of people, and differences between ways of thinking about fairness between the legal world (i.e., fairness as procedural) and the fair-ML community (i.e., fairness as outcome-based)."

What do folks think? @lgfunderburk @hildeweerts

Author: This section took me a bit more to develop and was recently pushed. The reword and new examples look as follows:

`This trap occurs when abstraction results in failure to take
into account the full picture behind fairness in a situation. Many tasks of a data
scientist involve some form of formalization: from measuring real-world phenomena as
data to translating business KPI's and constraints into metrics, loss functions, or parameters.
We fall into the formalism trap when this translation fails to account the social dimensions
associated to the problem. Social dimensions that cannot be abstracted through a purely mathematical
framework include procedurality, contextuality and contestability.

Example: Consider the use of "risk scores" to decide the outcomes individuals
will experience. Risk scores may be used to determine risk of recidivism in crime,
risk of failure within a new role during recruitment, or risk of failure in a
patient's chosen treatment. A key element within these algorithms is the
ratio of false and true positives, as well as and false and true negatives.
That is, how successful an algorithm is to correctly asses risk in an individual.

A more important distinction is the impact of a wrongful assignment (is an individual
incarcerated, vs is an individual hired).

Take the following risk assessment formula from Kleinberg et al. [#6]_. The problem
of risk assessment can be abstracted as follows:

1. Feature vectors: this represent data we know about a given person.

2. Groups: this represents information regarding belonging of a person. Ideally our
algorithm does not display preference with respect to the group a person
belongs to. Example of groups include gender, race, age, etc.

3. Risk assignment: this represents the probability that a person with given
features, belonging to a given group, possesses a risk to engage in
well defined behaviour, and that the risk calculated accurately depicts reality for
that given person. In other words, the algorithm has low false positive rates and low false
negative rates.

From a mathematical point of view, it is possible to represent each of these
through the use of vectors, booleans (true or false) and probability values.
Where mathematical abstraction encounters a limitation is when capturing
information regarding procedurality (for example, how  do judges and police officers
determine whether bail, counselling, probation, or incarceration is appropriate);
contextuality (different societies determine what constitutes an "unfair" outcome, furthermore
different groups determine what constitutes immoral discrimination, i.e. is it immoral
to hire an applicant whose primary language is not English, for a non-English speaking role, over
an applicant whose only spoken language is English?); and contestability (the definitions
of discrimination and unfairness are politically contested and change over time, how
to mathematically represent definitions of unfairness as they are contested over time?).
`

It ended up being quite wordy. Open to reducing content

Reviewer 2: For simplicity, you could start with some of that language I commented above:

"Because different definitions of fairness cannot be satisfied at the same time, you'll need to decide which definition to use. But there's no mathematical way to make that decision. And it might be that none of the definitions are appropriate for your situation.

Some reasons why people fall into this trap is because fairness is contextually-dependent, open to contestation by different groups of people, and differences between ways of thinking about fairness between the legal world (i.e., fairness as procedural) and the fair-ML community (i.e., fairness as outcome-based)."

Author: Adding this

Author: Ok, added

Example:

The Portability Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when we fail to understand how reusing a model or
algorithm that is designed for one specific social context, may not necessarily
apply to a different social context. Reusing an algorithmic solution and failing
to take into account differences in involved social contexts can result in misleading
results and potentially harmful consequences if the algorithm is used to determine the
fate of an individual.

Example: Reusing a machine learning algorithm used to screen job applications in the
nursing industry, for job applications in the information technology sector. An intuitive
yet important difference between both contexts is the difference in skills required to
succeed in both industries. A slightly more subtle difference is the demographic differences

> in gender attraction to each of the industries, resulting from wording in job postings,

Reviewer 2: Is there a different way to say "gender attraction"? Maybe something like "gendered differences for job applicants in different industries, differences which may result from..."?

Author: Reworded into 

> A slightly more subtle difference is the demographic differences
in number of genders who typically work in each of these industries, which may result from
wording in job postings, social constructs on gender and societal roles, and the male-female
ratio of successful applicants in each field.

social constructs on gender and societal roles, and the male-female ratio of successful
applicants in each field.

The Framing Trap
^^^^^^^^^^^^^^^^

This trap occurs when data scientists fail to consider the full picture surrounding
a particular social context when abstracting a social problem, and implementing an

> algorithm in which the outcome involves enforcing decisions that will impact a person

Reviewer 2: I'm not sure I totally understood this (sorry, may be just late-afternoon lack of caffeination :)) - are you saying this applies for any algorithm that will impact a person or group of people? Or only ones where data scientists ("we"?) fail to consider the full picture of the social context?

Author: Agreed it was a confusing sentence
Reworded as follows:

This trap occurs when we fail to consider the full picture surrounding
a particular social context when abstracting a social problem. Elements
involved include but are not limited to: the social landscape that is being
studied, characteristics of individuals or circumstances of the chosen
 situation, third parties involved along with their circumstances,
and the task that is being set out to abstract (i.e. calculating a risk score,
choosing between a pool of candidates, selecting an appropriate treatment, etc).

or group of people.

> The authors define three main frames:

Reviewer 2: I love that you included these definitions in here! I wonder whether we can introduce it a bit for readers, something like:

"To help us avoid drawing narrow boundaries of what is considered in scope for the problem, we might consider different "frames" to use. The first, narrowest frame, focuses only on modeling choices made when designing the algorithm..."

Author: Incorporated sentence

1. The algorithmic frame: these are choices made when abstracting a problem
in the form of representations (data) and labelling (outcome). In this frame
an algorithm is evaluated based on properties of output in relation to input.
Examples of these properties include accuracy and generalizability to data the model
did not train on. Fairness is typically not evaluated in this frame.

2. The data frame: this frame request to the content of input and output of the
algorithm ('data') whether in the form of abstractions, or outcomes. This additional

> frame allows us to question the inherent (un)fairness present in input and output data.

Reviewer 2: This might be an easy place to put in a reference or link to ways of doing this, like datasheets (either the [paper](https://arxiv.org/pdf/1803.09010) or a package for it)

Author: Added brief comment and reference

existing groups in that context, or how actors might change their behaviors to
game the model.

The Formalism Trap
^^^^^^^^^^^^^^^^^^

Many tasks of a data scientist involve some form of formalization: from
measuring real-world phenomena as data to translating business KPI's
and constraints into metrics, loss functions, or parameters. We fall into the
formalism trap when we fail to account for the full meaning of social
concepts like fairness. This occurs because there is no purely mathematical way to resolve
conflicting definitions of fairness. This is also because fairness is complex and
contested by social actors, and it cannot only be captured mathematically, but
needs to be understood procedurally (i.e., [insert explanation here]) and situated in social contexts.

Because different definitions of fairness cannot be satisfied at the same time,
we'll need to decide which definition to use. But there's no mathematical way to
make that decision. And it might be that none of the definitions are appropriate for our situation.

Some reasons why we fall into this trap is because fairness is context-dependent,

> open to contestation by different groups of people, and differences between ways of thinking

Author: Incorporated Hilde's paragraphs

about fairness between the legal world (i.e., fairness as procedural)
and the fair-ML community (i.e., fairness as outcome-based).

> Kleinberg et al. [#6]_ abstract the problem risk assessment via the use of

Author: Incorporated her paragraphs, removed suggested sentences

vectors to represent information about  person, boolean values to
represent group belonging, and risk assignment scores.

Where mathematical abstraction encounters a limitation is when capturing
information regarding procedurality (for example, how do judges and police officers
determine whether bail, counselling, probation, or incarceration is appropriate);
contextuality (different societies determine what constitutes an "unfair" outcome, furthermore
different groups determine what constitutes immoral discrimination, i.e. is it immoral
to hire an applicant whose primary language is not English, for a non-English speaking role, over

> an applicant whose only spoken language is English?); and contestability (the definitions

Author: rephrased

which are typically described by inputs, outputs, and the relationship between them.
However, by abstracting away the social context they are at risk of falling into
'abstraction traps': a failure to consider how social context and technology
are interrelated.

In this section, we explain what those traps are, and give some suggestions on
how we can avoid them.

In "Fairness and Abstraction in Sociotechnical Systems," Selbst et al. [#4]_
identify failure modes that can arise from abstracting away the social context
when modeling. They identify them as:

* *The Solutionism Trap*

* *The Ripple Effect Trap*

* *The Formalism Trap*

* *The Portability Trap*

* *The Framing Trap*

We provide some definitions and examples of these traps to help Fairlearn
users think about how choices they make in their work can lead to or avoid
these common pitfalls.

The Solutionism Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when we assume that the best solution to a problem
may involve technology, and fail to recognize other possible solutions
outside of this realm. Solutionist approaches may also not be appropriate
in situations where definitions of fairness may change over time

> (see :ref:`formalism_trap`).

Reviewer 2: It doesn't look like this ref is working properly, unless I'm missing something

Author: I believe this was fixed in a later commit via one of your suggestions

> (see :ref:`formalism_trap`).

Reviewer 2: ```suggestion
(see :ref:`The Formalism Trap`).
```

Reviewer 2: This should fix the reference!

Reviewer 3: I'm confused. It's defined below (currently line 222), why wouldn't it work?

Reviewer 2: Hmm, not sure. It's currently displaying as ":ref:`formalism_trap`" though

Reviewer: Huh, that's weird... Did anybody figure out yet why this doesn't work?

Author: Applied suggestion. Let me know if the problem persists

Example: consider the problem of internet connectivity in rural communities.

> An example of the solutionism trap is assuming that by using data science to

Reviewer 2: ```suggestion
An example of the solutionism trap is assuming that using data science to
```

> study internet speed in a given region, insights we gain from using data

Reviewer 2: ```suggestion
measure internet speed in a given region
```

can help improve internet connectivity.

>  However, if there are additional socioeconomic challenges within

Reviewer 2: ```suggestion
However, if there are additional socioeconomic challenges within
```

Reviewer 2: The initial space before "However" was making it render as an indented line

> science can help us in negotiating deals or discovering potential for good

Reviewer 2: ```suggestion
can help improve internet connectivity.
```

> policies. However, if there are additional socioeconomic challenges within

Reviewer 2: ```suggestion
 However, if there are additional socioeconomic challenges within
```

a community, for example with education, infrastructure, information
technology, or health services, then an algorithmic solution purely
focused on internet speed may fail to meaningfully address the needs of
the community.

The Ripple Effect Trap
^^^^^^^^^^^^^^^^^^^^^^

This trap occurs when we do not consider the unintended consequences of
introducing technology into an existing social system. Such consequences
include changes in behaviors, outcomes, individual experiences, or changes

> in underlying social values and incentives of a given social system, for

Reviewer 2: ```suggestion
in underlying social values and incentives of a given social system; for
```

> instance by increasing perceived value of quantifiable metrics over

Reviewer 2: ```suggestion
instance, by increasing perceived value of quantifiable metrics over
```

recommendations are good enough without those additional sources of information.

To avoid this pitfall, we must be aware that once a technology is incorporated
into a social context, new groups may reinterpret it differently. We should
adopt "what if" scenarios to envision how the social context might change
after introducing a model, including how it may change the power dynamics of
existing groups in that context, or how actors might change their behaviors to
game the model.

.. _<formalism_tag>:
The Formalism Trap
^^^^^^^^^^^^^^^^^^

Many tasks of a data scientist involve some form of formalization: from
measuring real-world phenomena as data to translating business KPI's
and constraints into metrics, loss functions, or parameters. We fall into the
formalism trap when we fail to account for the full meaning of social
concepts like fairness.

Fairness is a complex construct that is contested: different people may
have different ideas of what is fair in a particular scenario. While
mathematical fairness metrics may capture some aspects of fairness, they
fail to capture all relevant aspects. For example, group fairness metrics

> do not account for differences in individual experiences nor do they

Reviewer 2: ```suggestion
do not account for differences in individual experiences, nor do they
```

account for procedural justice.

In some scenarios, fairness metrics such as demographic parity and equalized
odds cannot be satisfied at the same time. At a first glance, this may appear
to be a mathematical problem. However, the conflict is actually grounded in
different understandings of what fairness is. Consequently, there is no
mathematical approach to solve the conflict. Instead we need to decide which
metrics might be appropriate for the situation at hand, keeping in mind the
limitations of a mathematical formalization. In some cases, there may be no
suitable metric.

Some reasons why we fall into this trap are because fairness is
context-dependent, because it is open to contestation by different groups
of people, and because there are differences between ways of thinking about
fairness between the legal world (i.e., fairness as procedural) and the fair-ML
community (i.e., fairness as outcome-based).

Where mathematical abstraction encounters a limitation is when

> capturing information regarding procedurality (for example, how do judges

Reviewer 2: Ah, I think I had left a comment previously about the sequencing between this paragraph and the previous one. 

The previous one is 123 (contextual; contested; procedural) and this one is 312 (procedural; contextual; contested)...

Can we make both of them the same sequence? That'll make it easier for readers to follow

Author: Reordered

and police officers determine whether bail, counselling, probation, or
incarceration is appropriate); contextuality (different communities may have
different definitions for what constitutes an "unfair" outcome;

>  for instance, is it

Reviewer 2: ```suggestion
for instance, is it
```

Reviewer 2: Same issue with indenting as the "However" above - this should fix it

> different definitions for what constitutes an "unfair" outcome; different

Reviewer 2: ```suggestion
different definitions for what constitutes an "unfair" outcome;
```

> groups determine what constitutes immoral discrimination, for instance, is it

Reviewer 2: ```suggestion
 for instance, is it
```

unfair to hire an applicant whose primary language is English, for an
English speaking role, over an applicant whose only spoken language is

> not English?); and contestability (the definitions of discrimination and

Reviewer 2: This paragraph is a bit difficult to read - I'd recommend either bolding or italicizing the 3 terms (procedurality, contextuality, and contestability), or move each of these into bullets.

@lgfunderburk can you also swap the order of the three in this paragraph to reflect the order you introduced them in in the previous paragraph (which is currently: contextual, contestable, and procedural)? Or, switch the order of the previous paragraph - either way, it'll make it easier for readers to follow if the order is consistent.

unfairness are politically contested and change over time, which may pose
fundamental challenges for representing them mathematically).

The Portability Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when we fail to understand how reusing a model or
algorithm that is designed for one specific social context may not
necessarily apply to a different social context. Reusing an algorithmic
solution and failing to take into account differences in involved social
contexts can result in misleading results and potentially harmful consequences.

For instance, reusing a machine learning algorithm used to screen

> job applications in the nursing industry, for a system used to screen

Reviewer 2: ```suggestion
job applications in the nursing industry for a system used to screen
```

Another key difference between these contexts involves the demographic
differences (in terms of gender) of employees in each of these industries,
which may result from wording in job postings, social constructs on gender
and societal roles, and the percentages of successful applicants in
each field per (gender) group.

The Framing Trap
^^^^^^^^^^^^^^^^

This trap occurs when we fail to consider the full picture surrounding
a particular social context when abstracting a social problem. Elements
involved include but are not limited to: the social landscape that the
chosen phenomenon exists in, characteristics of individuals or
circumstances of the chosen situation, third parties involved along with
their circumstances, and the task that is being set out to abstract
(i.e., calculating a risk score, choosing between a pool of candidates,
selecting an appropriate treatment, etc).

To help us avoid drawing narrow boundaries of what is considered in scope
for the problem, we might consider using wider "frames" around what is
considered to be in scope for the problem, moving from an algorithmic frame
to a sociotechnical frame.

> For instance, adopting a _sociotechnical_ frame (instead of a data-focused,

Reviewer 2: ```suggestion
For instance, adopting a *sociotechnical* frame (instead of a data-focused,
```

or algorithmic frame) allows us to recognize that a machine learning model
is part of social and technical interactions between people and technology,
and thus the social components of a given social context should be included
as part of the problem formulation and modeling approach (including local
decision-making processes, incentive structures, institutional processes,
and more).

For instance, we might fall into this trap by assessing risk of re-engagement
in criminal behavior for an individual charged with an offense, while failing
to consider factors such as the legacy of racial biases in criminal justice
systems, the relationship of socio-economic status and mental health to the
social construction of criminality, along with existing societal biases of
judges, police officers, or other social actors involved in the larger
sociotechnical frame around a criminal justice algorithm.

> Within the sociotechnical framework the model incorporates not only more

Reviewer 2: ```suggestion
Within the sociotechnical frame the model incorporates not only more
```

nuanced data on the history of the case, but also the social context in
which judging and recommending an outcome take place. This frame might
incorporate the processes associated with crime reporting, the offense-trial

> pipeline, and identify how the relationship between various social actors and

Reviewer 2: ```suggestion
pipeline, and an awareness of how the relationship between various social actors and
```

between the maximum and the minimum values. For more information refer to the
subpackage :code:`fairlearn.metrics`.

What traps can we fall into when modeling a social problem?
--------------------------------------------------------------

Machine learning systems used in the real world are inherently sociotechnical
systems, which include both technologies and social actors. Designers of machine
learning systems typically translate a real-world context into a machine learning
model through abstraction: focussing only on 'relevant' aspects of that context,
which are typically described by inputs, outputs, and the relationship between them.
However, by abstracting away the social context they are at risk of falling into
'abstraction traps': a failure to consider how social context and technology is interrelated.

In this section, we explain what those traps are, and give some suggestions on
how we can avoid them.

In "Fairness and Abstraction in Sociotechnical Systems," Selbst et al. [#4]_

> identify failure modes that can arise from abstracting away the social context when modeling. They identify

Reviewer 2: @lgfunderburk it's great that this is at this stage of comments! Means it's getting very close!! :)

them as:

* *The Solutionism Trap*

* *The Ripple Effect Trap*

* *The Formalism Trap*

* *The Portability Trap*

* *The Framing Trap*

We provide some definitions and examples of these traps to help Fairlearn
users think about how choices they make in their work can avoid these common pitfalls.

The Solutionism Trap
^^^^^^^^^^^^^^^^^^^^

This trap occurs when we assume that the best solution to a problem
may involve technology, and fail to recognize other possible solutions outside of
this realm. This trap also occurs as definitions of fairness are
contested and changed and models become obsolete. See the "Construct
Validity" section in Fairlearn's user guide.

Example: consider the problem of internet connectivity in rural communities.

> An example of the solutionism trap is assuming that by using data science to

Author: Added this clarification. Thank you!

study internet speed in a given region, policies and deals can be arranged to
reduce the price of internet access. However, if there are additional socioeconomic
problems within a community, for example lack of education, infrastructure, information
technology and health services, then an algorithmic solution purely focused on internet
speed will fail to meaningfully address the needs of the community.

The Ripple Effect Trap
^^^^^^^^^^^^^^^^^^^^^^

This trap occurs when we do not consider the unintended consequences of introducing
technology into an existing social system. Such consequences include change in
behaviour, outcomes individual experience and a change in underlying social values
and incentives of a given social system, for instance by increasing perceived value
of quantifiable metrics over non-quantifiable ones.

Example: consider the problem of banks deciding whether an individual should
be approved for a loan. Prior to using machine learning algorithms

> to compute a "score", banks might rely on engaging in conversations with

Reviewer 2: ```suggestion
to compute a "score", banks might rely on loan officers engaging in conversations with
```

Example: consider the problem of banks deciding whether an individual should
be approved for a loan. Prior to using machine learning algorithms

> to compute a "score", banks might rely on engaging in conversations with

Reviewer 2: Just added a suggested role to replace banks

> algorithm, it is possible that banks stop engaging in conversations

Reviewer 2: ```suggestion
algorithm, it is possible that loan officers stop engaging in conversations
```

clients, recommending a plan based on their unique situation, and
discussing with other team members to obtain feedback. By introducing an
algorithm, it is possible that banks stop engaging in conversations
with team members and clients, and assume the algorithm is good enough
to blindly accept a recommendation. Furthermore, some might use
information on people algorithm typically rejects to create a system
that takes advantage of people who need funds but did not receive them.

To avoid this pitfall, we must be aware that once a technology is incorporated

> into a social context, new groups may reinterpret it and potentially use this

Author: Thank you for suggestions. Applied changes.

information for outside purposes which include nefarious ones. To

> solve for this issue, we must incorporate "what if" scenarios into testing

Reviewer 2: ```suggestion
help address this issue, we should adopt "what if" scenarios to envision how the social context might change after introducing a model, including how it may change the power dynamics of existing groups in that context.
```

> solve for this issue, we must incorporate "what if" scenarios into testing

Author: Closing comment as it was solved with a suggestion, and suggestion was applied.

of the model.

The Formalism Trap
^^^^^^^^^^^^^^^^^^

This trap occurs when abstraction results in failure to take

> into account the full picture behind fairness in a situation. Many tasks of a data

Reviewer 2: I would just start with the "Many tasks" and remove the first sentence

Author: Done

> concepts like fairness. This occurs because there is no purely mathematical way to resolve

Reviewer: [if you have time]

This sentence suggests that we should somehow "resolve" conflicting definitions of fairness and that that's why we fall into the formalism trap. I think we can rephrase this paragraph a bit, e.g., something like:

"...concepts like fairness. 

Fairness is a complex construct that is contested: different people may have different ideas of what is fair in a particular scenario. While mathematical fairness metrics may capture some aspects of fairness, they fail to capture all relevant aspects. For example, group fairness metrics do not account for differences in individual experiences nor do they account for procedural justice.

In some scenarios, fairness metrics such as demographic parity and equalized odds cannot be satisfied at the same time. At a first glance, this may appear to be a mathematical problem. However, the conflict is actually grounded in different understandings of what fairness is. Consequently, there is no mathematical approach to solve the conflict. Instead we need to decide which metrics might are appropriate for the situation at hand, keeping in mind the limitations of a mathematical formalization. In some cases, there may be no suitable metric."

Reviewer 2: I like this suggested revision @hildeweerts!

Author: This works. 
Incorporated changes.

data to translating business KPI's and constraints into metrics, loss functions, or parameters.

> We fall into the formalism trap when this translation fails to account the social dimensions

Reviewer 2: ```suggestion
We fall into the formalism trap when we fail to account for the full meaning of social concepts like fairness. This occurs because there is no purely mathematical way to resolve conflicting definitions of fairness. This is also because fairness is complex and contested by social actors, and it cannot only be captured mathematically, but needs to be understood procedurally (i.e., [insert explanation here]) and situated in social contexts.
```

Reviewer 2: I think in a previous review I wrote a suggested version for this explanation (I'll see if I can find it and tag you in it). If it's easier, feel free to just use that language directly!

> associated to the problem. Social dimensions that cannot be abstracted through a purely mathematical

Reviewer 2: ```suggestion
```

> framework include procedurality, contextuality and contestability.

Reviewer 2: ```suggestion
```

Example: Consider the use of "risk scores" to decide the outcomes individuals
will experience. Risk scores may be used to determine risk of recidivism in crime,
risk of failure within a new role during recruitment, or risk of failure in a
patient's chosen treatment. A key element within these algorithms is the
ratio of false and true positives, as well as and false and true negatives.

> That is, how successful an algorithm is to correctly asses risk in an individual.

Reviewer 2: Hmm, I'm not sure this example gets this idea across.. @hildeweerts or @romanlutz thoughts?

Author: Removed example for now. Can create a different issue to select better examples as needed.

A more important distinction is the impact of a wrongful assignment (is an individual
incarcerated, vs is an individual hired).

> Take the following risk assessment formula from Kleinberg et al. [#6]_. The problem

Reviewer 2: Could just remove this example (including the 1, 2, and 3 below), and just weaving the examples below into the explanation above. To Miro's point, let's try to get this approved :) and future iterations can expand on it

> Take the following risk assessment formula from Kleinberg et al. [#6]_. The problem

Author: removed

an algorithm is evaluated based on properties of output in relation to input.
Examples of these properties include accuracy and generalizability to data the model
did not train on. Fairness is typically not evaluated in this frame.

2. The data frame: this frame request to the content of input and output of the
algorithm ('data') whether in the form of abstractions, or outcomes. This additional
frame allows us to question the inherent (un)fairness present in input and output data.

3. The sociotechnical frame: this frame recognizes that a machine learning model is part
of the interaction between people and technology, and thus any social components of this
interaction need to be modelled and incorporated accordingly.

Example: assessing risk of re-engagement in criminal behaviour in an individual
charged with an offense, and appropriate measures to prevent relapse, and failing
to consider factors such as race, socio-economic status, mental health, along with
socially-dependent views present in judges, police officers, or any actors responsible
for recommending a course of action.

In the algorithmic framework, for example, input variables may contain previous criminal history,

> statements taken by the accused, witnesses and police officers. Labels (outcome)

Reviewer 3: Yes, the algorithmic frame only includes the algorithm, not even the data.

> statements taken by the accused, witnesses and police officers. Labels (outcome)

Reviewer: That depends a bit on your definition of algorithm, I suppose. I'm not sure if we can really separate data/algorithm in the case of ML. You could view an ML model as a representation of the data it was trained on and in many cases an algorithm-centric approach towards fairness will depend on data at least to some extend (e.g., demographic data).

> statements taken by the accused, witnesses and police officers. Labels (outcome)

Reviewer 2: (Since we're recommending that @lgfunderburk remove this paragraph anyway, I'll resolve these comments.

include recommendations by the algorithm on an appropriate course of action based
on a computed risk score. Model is limited in assessing fairness out outcome.

The data framework could attempt to reduce unfairness by studying socio-economic

> information regarding the accused, their upbringing and how it relates to their

Reviewer 2: For simplicity to get this approved, we could remove the two paragraphs "In the algorithmic framework" and "The data framework could..." and only keep the sociotechnical framework paragraph below (since that's what we're hoping people will take anyway)

> information regarding the accused, their upbringing and how it relates to their

Reviewer 3: I agree with @mmadaio . I don't think we should encourage people to look at anything but the sociotechnical frame because it suggests that it's sufficient to do so.

> information regarding the accused, their upbringing and how it relates to their

Reviewer: Sounds good!

> information regarding the accused, their upbringing and how it relates to their

Author: Removed paragraphs

between the maximum and the minimum values. For more information refer to the
subpackage :code:`fairlearn.metrics`.

Failure modes encountered when abstracting
--------------------------------------------------

> In "Fairness and Abstraction in Sociotechnical Systems" Selbst et al. [#4]_ identify failure modes that arise the applying abstraction. They identify them as:

Reviewer 3: Did you mean "that arise from applying abstraction"?

Author: Yes! Solved

* *The Framing Trap*

* *The Portability Trap*

* *The Formalism Trap*

* *The Ripple Effect Trap*

* *The Solutionism Trap*

These traps are inspired by the observation that each of these traps are the result of failing to consider the way in with social context and technology are interrelated, as well as a deeper understanding of "the social" in order to solve problems Ackerman [#5]_. 

The Framing Trap
^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#5]_ define this as a "failure to model the entire system over which a social criterion, such as fairness, will be enforced".

Examples include:

> Labelling in a risk assessment setting. Risk assessment may be used as a predictive model whose goal is to assess the risk of an individual to ease the decision-making process. For example, using risk assessment to decide whether a person is eligible for morgtage approval. Another example is using risk assessment to decide whether a person is likely to re-engage in criminal behaviour as they move through the criminal justice pipeline. 

Reviewer 3: ```suggestion
Labelling in a risk assessment setting. Risk assessment may be used as a predictive model whose goal is to assess the risk of an individual to ease the decision-making process. For example, using risk assessment to decide whether a person is eligible for mortgage approval. Another example is using risk assessment to decide whether a person is likely to re-engage in criminal behavior as they move through the criminal justice pipeline. 
```

Author: Only noticed this after fixing typos. They appear to be fixed. let me know if they still pop up.

> As far as the outcomes of these decisions, in the first scenario an individual may be approved or not to get a mortgage which then impacts their ability to have housing security, gain wealth and social status. In the second scenario and an individual may be released pretrial, or bail may be required. This assessment may also include sentencing and parole hearing. 

Reviewer 3: ```suggestion
As far as the outcomes of these decisions, in the first scenario an individual may be approved or not to get a mortgage which then impacts their ability to have housing security, gain wealth and social status. In the second scenario an individual may be released pretrial, or bail may be required. This assessment may also include sentencing and parole hearing. 
```

between the maximum and the minimum values. For more information refer to the
subpackage :code:`fairlearn.metrics`.

> Failure modes encountered when abstracting

Author: Changed into 

> What traps can we fall into when abstracting a social problem?

--------------------------------------------------

> 

Author: modified to

> What traps can we fall into when abstracting a social problem?

> 

Reviewer: I was thinking we can add a sentence after the first one, e.g., something like:

"Machine learning systems used in the real world are inherently sociotechnical
 systems, which include both technologies and social actors. Designers of machine learning systems typically translate a real-world context into a machine learning model through abstraction: focussing only on 'relevant' aspects of that context, which are typically described by inputs, outputs, and the relationship between them. However, by abstracting away the social context they are at risk of falling into 'abstraction traps': a failure to consider how social context and technology is interrelated."

or something like that.

I personally prefer to talk from "we" perspective rather than "they", because it sounds a bit more humble and people might get less "computer science is bad and you should feel bad" vibes. But this is obviously a stylistic choice and I'm happy either way.

WDYT @mmadaio

> 

Author: Incorporated and reworded from "they" to "we"

In "Fairness and Abstraction in Sociotechnical Systems" Selbst et al. [#4]_

> identify failure modes that arise from applying abstraction. They identify

Author: Modified section

them as:

> * *The Framing Trap*

Author: reversed order of appearance

* *The Portability Trap*

* *The Formalism Trap*

* *The Ripple Effect Trap*

* *The Solutionism Trap*

> These traps are inspired by the observation that each of these traps are the

Author: Reworded sentence to

> This trap occurs when data scientists assume that the best solution to a problem
involves technology, and fail to recognize other possible solutions outside of
this realm

result of failing to consider the way in with social context and technology
are interrelated, as well as a deeper understanding of "the social" in order to
solve problems Ackerman [#5]_.

The Framing Trap
^^^^^^^^^^^^^^^^

> Selbst et al. [#4]_ define this as a "failure to model the entire system over

Author: Agreed. Removing citation from each section and adding explained definition instead.

> which a social criterion, such as fairness, will be enforced".

Reviewer 2: This might need a bit more explanation to be understandable for data scientists - maybe discuss the different frames they introduce and how that may shape fairness work?

Author: agreed, reworded

It is the practice of failing to consider the full picture surrounding sources of

> bias.

Reviewer 2: Ah, we recommend not using the term bias (since it's overloaded and means different things to different communities), and instead talking about fairness issues or fairness-related harms (see contributor guide section [here](https://fairlearn.org/v0.6.2/contributor_guide/how_to_talk_about_fairness.html))

Author: Removed word "bias" from definitions and examples, substituted into "social context", "social construct", or other context/social related sources of understanding.

> Labelling is a risk assessment setting. Risk assessment may be used as a predictive

Reviewer 2: Ah, I'm not sure I understand what that first sentence means...

Author: Reworded

model whose goal is to assess the risk of an individual to ease the decision-making
process. For example, using risk assessment to decide whether a person is eligible
for mortgage approval. Another example is using risk assessment to decide whether a
person is likely to re-engage in criminal behaviour as they move through the criminal
justice pipeline.

As far as the outcomes of these decisions go, in the first scenario an individual may be
approved or not to get a mortgage which then impacts their ability to have housing security,
gain wealth and social status. In the second scenario an individual may be released pretrial,
or bail may be required. This assessment may also include sentencing and parole hearing.
Risk assessment tools may be used as recommendations towards an outcome. However there is no
guarantee that they will be used consistently and thus bias is introduced. For example, a
party responsible for choosing the outcome decides to ignore the recommendation as a result
of personal bias, even when the model takes into account demographic data. In this example

> a model that fails to take into account bias present in those judging culminates in a

Reviewer 2: I like the use of examples here! It could be good to just focus on one example, though, and explicitly connect it back to the different types of frames (e.g., algorithmic, data, and sociotechnical) and why the sociotechnical frame matters. I think this is close! But it would be good to be more explicit about the connection between the example and the frame.

Reviewer 2: I think this one is really important to be clear about, since I think much of the value proposition of Fairlearn is that we encourage (and ideally provide support for!) fair-ML practitioners trying to shift from an algorithmic and/or data frame towards a sociotechnical frame

Author: Incorporated frameworks and connected single example to sociotechnical framework

biased result.

The Portability Trap
^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to understand how repurposing algorithmic
solutions designed for one social context may be misleading, inaccurate, or otherwise
do harm when applied to a different context".

> It is the practice of reusing others work without taking into account lack of

Reviewer 2: Hmm, I'm not sure that it's necessarily just re-using others' work, but crucially that piece about transferring from one social context to another. And one other nuance they offer in the paper is this doesn't only refer to domain (e.g., from hiring to criminal justice) but also locales (e.g., geographic or sociocultural contexts), which they give the example of different court jurisdictions and why that matters

Reviewer 2: I think an example could be really helpful here in demonstrating why this matters and what impact (not) considering this might have

transferability to a given social situation, resulting in harmful consequences.

The Formalism Trap
^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to account for the full meaning of social concepts
such as fairness, which can be procedural, contextual, and contestable, and cannot be resolved

> through mathematical formalisms".

Reviewer 2: I think this is going to be a really interesting one to think through what it means for users of Fairlearn (I know this isn't a particularly helpful or actionable comment, but just wanted to flag this for discussion at some point :) )

Author: Reworded for now
added example, but example likely needs more work

It is the practice of implementing mathematical and statistical models, along with
corresponding assumptions, that fail to take into consideration the social, demographic,

> economic, or otherwise non-technical aspect that make up the phenomenon being studied.

Reviewer 2: An example here would also be really helpful. 

And I think we should probably briefly discuss the two problems they discuss with formalisms, e.g.,
- in the face of irreconcilably contested definitions, social context (and engagement with stakeholders from that context!) should help inform decisions about which definition to use
- mathematical formalisms abstract away the procedural, contextual, and contestable nuances of fairness

Author: added examples and expanded this section a lot more, mentioned how mathematical formalists are limited when it comes to representing procedural, contextual and contestable nuances of fairness aspects

The Ripple Effect Trap
^^^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to understand how the insertion of technology into
an existing social system changes the behaviors and embedded values of the pre-existing system".

> 

Reviewer 2: This one is a bit more straightforward than the formalism trap, but a somewhat more extended definition may be helpful! (e.g., something that gets at the fact that ML systems may change _people's_ behaviors, organizations' incentives, or other social forces or actors).

And, as with the previous two, an example would be super helpful here!

Author: Added similar example, but substituted for loans as I mention the judge one in the next section

The Solutionism Trap
^^^^^^^^^^^^^^^^^^^^

Selbst et al. [#4]_ define this as a "failure to recognize the possibility that the best solution
to a problem may not involve technology".

> It is the practice of assuming that a machine learning algorithm is the best solution to a problem.

Reviewer 2: Should this be "...algorithm *may not be* the best solution..."?

Author: Resolved

One area where this manifests in contexts in which the definition "fairness" changes or is dependent

> on a political context. When this happens, models become obsolete. Another area where this manifests

Reviewer 2: I would argue (and I think Selbst et al do too) that all definitions of fairness are political :) Perhaps the relevant aspect of this is when they're politically _contested_?

Reviewer 2: This is also something that @michaelamoako has been thinking about with the construct validity glossary (PR #787)! Would be good to put in a pointer to that whenever it's ready

Author: This is an awesome pointer! How can we incorporate it?

Reviewer 2: His PR is in the user guide now! Woo!!

https://fairlearn.org/v0.7.0/user_guide/fairness_in_machine_learning.html#what-is-construct-validity

Author: I added a note 

`See the "Construct Validity" section in Fairlearn's user guide.`

But unsure how to add a pointer.

> is when the question at hand requires a computationally intractable solution.

Reviewer 2: An example would help make this piece clearer as well

Author: added example

This trap occurs when we assume that the best solution to a problem
may involve technology, and fail to recognize other possible solutions outside of
this realm.  Solutionist approaches may also not be appropriate in situations
where definitions of fairness may change over time (see the Formalism Trap below for more).

> See also the "Construct Validity" section in Fairlearn's user guide.

Reviewer 3: For linking to other sections (and traps) we can first tag those sections
```
.. _<tag>:
<section header>
<section>

```
and then reference it
```
:ref:`<tag>`
```
I'm sure you'll find plenty of examples in the docs (including this very file). 🙂

   .. [#3] Agarwal, Beygelzimer, Dudik, Langford, Wallach `"A Reductions
      Approach to Fair Classification"
      <https://arxiv.org/pdf/1803.02453.pdf>`_, ICML, 2018.
	  

>    .. [#4] Selbst, Andrew D. and Boyd, Danah and Friedler, Sorelle and Venkatasubramanian, Suresh and Vertesi, Janet, "Fairness and Abstraction in     	  		Sociotechnical Systems" (August 23, 2018). 2019 ACM Conference on Fairness, Accountability, and Transparency (FAT*), 59-68, Available at `SSRN: 		<https://ssrn.com/abstract=3265913>`_, 

Reviewer 3: Something is a little off about the line wrap here. Let's try to stay below 80 characters, unless it would interrupt a link/URL.

Same below, and you can continue the following line right below the `[#5]` the same way it's been done above with `[#3]`.

Author: Tried shortening it in my latest commit. Let me know if this works better.

Reviewer 3: The second, third, etc. lines are still indented by 2 characters more than they'd need to be indented, but now I'm really nitpicking 😆 Sorry, I'll restrict myself to comments on the actual substance for now.

Author: Can you please confirm if the spacing looks more like what you had in mind @romanlutz ?

Reviewer 3: Yes, looks great! Thank you @lgfunderburk!

   .. [#3] Agarwal, Beygelzimer, Dudik, Langford, Wallach `"A Reductions
      Approach to Fair Classification"
      <https://arxiv.org/pdf/1803.02453.pdf>`_, ICML, 2018.
	  
   .. [#4] Selbst, Andrew D. and Boyd, Danah and Friedler, Sorelle and Venkatasubramanian,
      Suresh and Vertesi, Janet, "Fairness and Abstraction in Sociotechnical Systems"
      (August 23, 2018). 2019 ACM Conference on Fairness, Accountability, and Transparency
      (FAT*), 59-68, Available at `SSRN: 	<https://ssrn.com/abstract=3265913>`_,
   
   .. [#5] Mark S. Ackerman. 2000. `"The intellectual challenge of CSCW: The gap between
      social requirements and technical feasibility"
      https://doi.org/10.1207/S15327051HCI1523_5`. Human-Computer Interaction 15,
      2-3 (2000), 179–203.

   .. [#6] Jon Kleinberg, Sendhil Mullainathan, and Manish Raghavan. 2017.
      `"Inherent trade-offs in the fair determination of risk scores"
      https://arxiv.org/abs/1609.05807`. In Proc. of ITCS.

   .. [#7] Timnit Gebru, Jamie Morgenstern, Briana Vecchione, Jennifer
      Wortman Vaughan, Hanna Wallach, Hal Daumé III, Kate Crawford.
      "Datasheets for Datasets" `https://arxiv.org/abs/1803.09010

>       <https://arxiv.org/abs/1803.09010>`_ (2018)

Reviewer 3: Why are these included here? I don't have anything against them (I've only read 6 and 7), but they're not mentioned in the text above. You could add them to this section, though: https://fairlearn.org/main/user_guide/further_resources.html#papers

   .. [#3] Agarwal, Beygelzimer, Dudik, Langford, Wallach `"A Reductions
      Approach to Fair Classification"
      <https://arxiv.org/pdf/1803.02453.pdf>`_, ICML, 2018.
	  
   .. [#4] Selbst, Andrew D. and Boyd, Danah and Friedler, Sorelle and Venkatasubramanian,
      Suresh and Vertesi, Janet, "Fairness and Abstraction in Sociotechnical Systems" (August 23, 2018).
      2019 ACM Conference on Fairness, Accountability, and Transparency (FAT*), 59-68, Available at
      `SSRN: 	<https://ssrn.com/abstract=3265913>`_,
   

>    .. [#5] Mark S. Ackerman. 2000. The intellectual challenge of CSCW: The gap between social requirements

Reviewer 3: nit: can we add a link for these last two? Either from the venue/journal they were submitted to, or from ArXiv

Author: Added link

Reviewer 2: The CSCW nerd in me is thrilled that Ackerman is cited in Fairlearn!! :)

   .. [#3] Agarwal, Beygelzimer, Dudik, Langford, Wallach `"A Reductions
      Approach to Fair Classification"
      <https://arxiv.org/pdf/1803.02453.pdf>`_, ICML, 2018.
	  
   .. [#4] Selbst, Andrew D. and Boyd, Danah and Friedler, Sorelle and Venkatasubramanian,
      Suresh and Vertesi, Janet, "Fairness and Abstraction in Sociotechnical Systems" (August 23, 2018).
      2019 ACM Conference on Fairness, Accountability, and Transparency (FAT*), 59-68, Available at
      `SSRN: 	<https://ssrn.com/abstract=3265913>`_,
   
   .. [#5] Mark S. Ackerman. 2000. `"The intellectual challenge of CSCW: The gap between social requirements
      and technical feasibility" https://doi.org/10.1207/S15327051HCI1523_5`. Human-Computer
      Interaction 15, 2-3 (2000), 179–203.

   .. [#6] Jon Kleinberg, Sendhil Mullainathan, and Manish Raghavan. 2017. `"Inherent trade-offs in the fair
      determination of risk scores" https://arxiv.org/abs/1609.05807`. In Proc. of ITCS.

   .. [#7] Timnit Gebru, Jamie Morgenstern, Briana Vecchione, Jennifer Wortman Vaughan,

>       Hanna Wallach, Hal Daumé III, Kate Crawford. `"Datasheets for Datasets" https://arxiv.org/abs/1803.09010`

Reviewer 3: Should we make these links into actual links?
```suggestion
      Hanna Wallach, Hal Daumé III, Kate Crawford. "Datasheets for Datasets" `https://arxiv.org/abs/1803.09010 <https://arxiv.org/abs/1803.09010>`_
```
